algebra-core/src/fields/models/fp4.rs

```diff
@@ -0,0 +1,427 @@
+    }
+}
+
+impl<P: Fp4Parameters> From<u16> for Fp4<P> {
+    fn from(other: u16) -> Self {
+        Self::new(other.into(), <Fp2<P::Fp2Params>>::zero())
+    }
+}
+
+impl<P: Fp4Parameters> From<u8> for Fp4<P> {
+    fn from(other: u8) -> Self {
+        Self::new(other.into(), <Fp2<P::Fp2Params>>::zero())
+    }
+}
+
+impl<P: Fp4Parameters> ToBytes for Fp4<P> {
+    #[inline]
+    fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        self.c0.write(&mut writer)?;
+        self.c1.write(writer)
+    }
+}
+
+impl<P: Fp4Parameters> FromBytes for Fp4<P> {
+    #[inline]
+    fn read<R: Read>(mut reader: R) -> IoResult<Self> {
+        let c0 = <Fp2<P::Fp2Params>>::read(&mut reader)?;
+        let c1 = <Fp2<P::Fp2Params>>::read(reader)?;
+        Ok(Fp4::new(c0, c1))
+    }
+}
+
+impl<P: Fp4Parameters> Neg for Fp4<P> {
+    type Output = Self;
+    #[inline]
+    #[must_use]
+    fn neg(self) -> Self {
+        let mut res = self.clone();
+        res.c0 = res.c0.neg();
+        res.c1 = res.c1.neg();
+        res
+    }
+}
+
+impl<P: Fp4Parameters> Distribution<Fp4<P>> for Standard {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Fp4<P> {
+        Fp4::new(UniformRand::rand(rng), UniformRand::rand(rng))
+    }
+}
+
+impl<'a, P: Fp4Parameters> Add<&'a Fp4<P>> for Fp4<P> {
+    type Output = Self;
+
+    #[inline]
+    fn add(self, other: &Self) -> Self {
+        let mut result = self;
+        result.add_assign(other);
+        result
+    }
+}
+
+impl<'a, P: Fp4Parameters> Sub<&'a Fp4<P>> for Fp4<P> {
+    type Output = Self;
+
+    #[inline]
+    fn sub(self, other: &Self) -> Self {
+        let mut result = self;
+        result.sub_assign(other);
+        result
+    }
+}
+
+impl<'a, P: Fp4Parameters> Mul<&'a Fp4<P>> for Fp4<P> {
+    type Output = Self;
+
+    #[inline]
+    fn mul(self, other: &Self) -> Self {
+        let mut result = self;
+        result.mul_assign(other);
+        result
+    }
+}
+
+impl<'a, P: Fp4Parameters> Div<&'a Fp4<P>> for Fp4<P> {
+    type Output = Self;
+
+    #[inline]
+    fn div(self, other: &Self) -> Self {
+        let mut result = self;
+        result.mul_assign(&other.inverse().unwrap());
+        result
+    }
+}
+
+impl<'a, P: Fp4Parameters> AddAssign<&'a Self> for Fp4<P> {
+    #[inline]
+    fn add_assign(&mut self, other: &Self) {
+        self.c0.add_assign(&other.c0);
+        self.c1.add_assign(&other.c1);
+    }
+}
+
+impl<'a, P: Fp4Parameters> SubAssign<&'a Self> for Fp4<P> {
+    #[inline]
+    fn sub_assign(&mut self, other: &Self) {
+        self.c0.sub_assign(&other.c0);
+        self.c1.sub_assign(&other.c1);
+    }
+}
+
+impl_additive_ops_from_ref!(Fp4, Fp4Parameters);
+impl_multiplicative_ops_from_ref!(Fp4, Fp4Parameters);
+
+impl<'a, P: Fp4Parameters> MulAssign<&'a Self> for Fp4<P> {
+    #[inline]
+    fn mul_assign(&mut self, other: &Self) {
+        // Karatsuba multiplication;
+        // Guide to Pairing-based cryprography, Algorithm 5.16.
```

Comment:
Maybe also add a reference to a publicly available resource, like https://github.com/scipr-lab/zexe/blob/4ab4b17ecec7ac6c635b4873321f7015a308273a/r1cs-std/src/fields/fp2.rs#L183